Ci/oe selftest

lbonn opened the pull request:
Based on #534.

lbonn commented:
Ok it's been quite chaotic, I'll squash all of that at some point.
But it seems that it's getting close to the end.

patrickvacek commented on lib/oeqa/selftest/cases/testutils.py:
```diff
@@ -60,32 +61,38 @@ def qemu_bake_image(imagename):
-    testInst.assertTrue(os.path.isfile(p), msg="No {} found ({})".format(program, p))
-    env = dict(os.environ)
-    env['LD_LIBRARY_PATH'] = libdir
-    result = runCmd(cmd, env=env, native_sysroot=sysroot, ignore_status=True, **kwargs)
```
Are you sure it's safe to remove the LD_LIBRARY_PATH bit? I seem to recall that was necessary for some obscure reason. If only I'd left a comment...

lbonn replied:
I actually had problem with .so before this change. It looks like we had to do that because the library path was actually outside of the sysroot. With `run_cmd(..., native_sysroot=...)` and the bitbake trick from there https://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/lib/oeqa/selftest/cases/package.py?id=2a6094ca8c4aa5c00636df77669e5bca6ad52bcf#n29, it seems that it works with less confusion.

Might need a double check on rocko and others though.

patrickvacek replied:
Cool, no objection if it works!

patrickvacek commented on lib/oeqa/selftest/cases/updater_qemux86_64.py:
```diff
@@ -39,20 +41,18 @@ def setUpClass(cls):
     def test_cert_provider_local_output(self):
-        logger = logging.getLogger("selftest")
-        logger.info('Running bitbake to build aktualizr-device-prov')
-        bitbake('aktualizr-device-prov')
```
So are you expecting bitbake to have been run first? Probably okay, but causes problems if you try to run this test in isolation or in an unexpected sequence.

lbonn replied:
I moved it to `setUpClass()`, so I guess it should always happen, no?

patrickvacek replied:
Oops, yes of course. Sorry, using github on a phone is difficult.

patrickvacek commented on scripts/ci/Dockerfile.bitbake:
```diff
@@ -1,42 +1,55 @@
-ARG uid=1000
-ARG gid=1000
+ARG uid=4321
+ARG gid=4321
```
What's the deal with this special user?

lbonn replied:
That was the uid/gid of Jenkins and the permissions of the bitbake cache matches that.

patrickvacek commented on lib/oeqa/selftest/cases/testutils.py:
```diff
@@ -60,32 +61,38 @@ def qemu_bake_image(imagename):
+    path = os.path.abspath(os.path.dirname(__file__))
+    metadir = path + "/../../../../../"
+
+    return metadir
```
Good call to pull this out.

lbonn commented:
Did some cleanup:

- added comments about oe-selftests skipping ostreepush
- added comments on the checkout step
- added comments on the kvm trick
- squashed some similar commits

lbonn commented:
Ok, I've made some more changes. It should have enough features to be used on other yocto layers + aktualizr and pin some versions just in case.

patrickvacek commented on lib/oeqa/selftest/cases/updater_qemux86_64.py:
```diff
@@ -263,18 +263,17 @@ def test_provisioning(self):
-            if p11_ret == 0 and hsm_ret == 0 and hsm_err == b'':
-                ran_ok = True
+            if (p11_ret == 0 and hsm_ret == 0 and hsm_err == b'' and
+                    b'X.509 cert' in p11_out and b'present token' in p11_err):
```
Good call.

lbonn replied:
Yes, I think it was the cause of some random failures. Haven't seen any since then.

patrickvacek commented on scripts/ci/gitlab/docker.yml:
```diff
@@ -20,6 +22,8 @@ Docker Setup (local):
     - docker push "$CHECKOUT_IMAGE"
 
 Docker Setup (simple):
+  # intended to be un on other branches and repos: just pulls the last master
```
"un on"?

lbonn replied:
thanks!

lbonn commented:
Ok the scripts have been changed significantly again and use [the extends feature from gitlab](https://docs.gitlab.com/ee/ci/yaml/#extends) instead of less-powerful yaml anchors.

I'll change the [rocko example PR](https://github.com/advancedtelematic/meta-updater/pull/535) to reflect these changes and start the integration with aktualizr for ptest.

patrickvacek commented:
Looks cool, but I only see the qemux86-64 tests running. Should the others run too?

lbonn commented:
I've only enabled the qemu tests for now. As we have two build machines, minnowboard and raspberrypi would lengthen the tests some more (maybe double the time) and they tend to break even less often than the rest.

Maybe as nightly jobs?

patrickvacek commented:
> Maybe as nightly jobs?

That's fine for me, we can punt on that for the moment. But I didn't see the native tests, either.

lbonn commented:
Ah my bad sorry, that's indeed a mistake. Thanks!

lbonn commented:
ptest is far from being green but I think this is finally ready now
